### PR TITLE
Add Factory method in BusManager.

### PR DESCRIPTION
**Pull request by siarheishut**
Tested with unit-tests ans manually.

### CODE REVIEW

File: tests/bus_manager_test.cpp [hunk 1]
```diff
@@ -1,11 +1,171 @@
...
+                  .stops = {},
+              },
+          },
+          .want = false,
+      },
+      TestCase{
+          .name = "Unknown stop along the route",
+          .config = {
+              PostBusRequest{
+                  .bus = "stop 1",
```
Comment (foxfrikses) [inline tests/bus_manager_test.cpp:79]: rename bus

File: tests/bus_manager_test.cpp [hunk 2]
```diff
@@ -1,11 +1,171 @@
...
+                  .stops = {"stop 1", "stop 2", "stop 4"},
```
Comment (foxfrikses) [inline tests/bus_manager_test.cpp:80]: Is this a valid route?

File: tests/bus_manager_test.cpp [hunk 3]
```diff
@@ -1,11 +1,171 @@
...
+                  .stops = {{"stop 4", 5000}},
+              },
+              PostStopRequest{
+                  .stop = "stop 3",
+                  .coords = {33.333333, 33.333333},
+                  .stops = {{"stop 1", 4000}},
+              },
+          },
+          .want = false,
+      },
```
Comment (foxfrikses) [inline tests/bus_manager_test.cpp:124]: Add testcase for 'unused stop': add stop happened, but the stop wasn't used in any route or dist

File: tests/bus_manager_test.cpp [hunk 4]
```diff
@@ -1,11 +1,171 @@
...
+          .want = true,
+      },
+  };
+
+  for (auto &[name, config, want] : test_cases) {
+    if (!BusManager::Create(config)) {
+      EXPECT_FALSE(want);
+      continue;
+    }
+    EXPECT_TRUE(want);
```
Comment (foxfrikses) [inline tests/bus_manager_test.cpp:165]:
```
bool got = BusManager::Create(config).has_value();
EXPECT_EQ(want, got) << name;
```

File: src/bus_manager.cpp [hunk 5]
```diff
@@ -37,6 +37,35 @@ int ComputeUniqueCount(std::vector<std::string> stops) {
...
 
 namespace rm {
+std::unique_ptr<BusManager> BusManager::Create(std::vector<PostRequest> &requests) {
+  std::set<std::string> route_stops, road_dists_stops, stop_requests_stops;
+  std::set<std::string> buses;
+  for (auto &req : requests) {
+    if (auto ptr_b = std::get_if<PostBusRequest>(&req)) {
+      for (auto &stop : ptr_b->stops) {
+        route_stops.insert(stop);
+      }
```
[the remark this answers is not in the document]
Comment (siarheishut) [reply src/bus_manager.cpp:47]: Done.

File: src/bus_manager.cpp [hunk 6]
```diff
@@ -37,6 +37,35 @@ int ComputeUniqueCount(std::vector<std::string> stops) {
...
+      auto [_, inserted] = buses.insert(ptr_b->bus);
+      if (!inserted) return nullptr;
+    } else if (auto ptr_s = std::get_if<PostStopRequest>(&req)) {
+      for (auto &[dst_stop, dist] : ptr_s->stops) {
+        road_dists_stops.insert(dst_stop);
+      }
+      auto [_, inserted] = stop_requests_stops.insert(ptr_s->stop);
+      if (!inserted) return nullptr;
```
[the remark this answers is not in the document]
Comment (siarheishut) [reply src/bus_manager.cpp:55]: Done.

File: src/bus_manager.cpp [hunk 7]
```diff
@@ -37,6 +37,32 @@ int ComputeUniqueCount(std::vector<std::string> stops) {
 }
 
 namespace rm {
+std::unique_ptr<BusManager> BusManager::Create(std::vector<PostRequest> &requests) {
+  std::set<std::string> route_stops, road_dists_stops, stop_requests_stops;
+  std::set<std::string> buses;
```
Comment (foxfrikses) [inline src/bus_manager.cpp:42]: Do you need it to be `std::string`, or can you use `std::string_view`?

File: tests/bus_manager_test.cpp [hunk 8]
```diff
@@ -74,10 +366,11 @@ TEST(TestBusManager, TestGetBusInfo) {
   };
 
   for (auto &[name, test_item, requests, want] : test_cases) {
-    BusManager bm(test_item);
+    auto bm = BusManager::Create(test_item);
+    EXPECT_TRUE(bm) << name;
```
Comment (foxfrikses) [inline tests/bus_manager_test.cpp:370]: If it's not true, you program goes on executing `bm->GetBusInfo` and so on, dereferencing nullptr. So, add `if (!bm) continue;` after `EXPECT_TRUE` to ensure this never happens.

File: tests/bus_manager_test.cpp [hunk 9]
```diff
@@ -165,13 +458,14 @@ TEST(TestBusManager, TestGetStopInfo) {
...
```
Comment (foxfrikses) [inline tests/bus_manager_test.cpp:462]: Same as above.

File: tests/bus_manager_test.cpp [hunk 10]
```diff
@@ -165,13 +458,14 @@ TEST(TestBusManager, TestGetStopInfo) {
...
 
     for (int i = 0; i < requests.size(); ++i) {
-      auto got = bm.GetStopInfo(requests[i].stop);
+      auto got = bm->GetStopInfo(requests[i].stop);
 
       if (!want[i].has_value()) {
-        EXPECT_TRUE(!got.has_value()) << name;
+        EXPECT_FALSE(got.has_value()) << name;
         continue;
       }
```
Comment (foxfrikses) [inline tests/bus_manager_test.cpp:470]:
```
EXPECT_EQ(want[i].has_value(), got.has_value()) << name;
if (!got) continue;
```

File: tests/bus_manager_test.cpp [hunk 11]
```diff
@@ -6,6 +6,298 @@
...
+                  .stops = {},
+              },
+          },
+          .want = true,
+      },
+  };
+
+  for (auto &[name, config, want] : test_cases) {
+    bool got = true;
+    if (!BusManager::Create(config)) got = false;
```
Comment (foxfrikses) [inline tests/bus_manager_test.cpp:296]: Why not `bool got = BusManager::Create(config) != nullptr;`?

File: tests/bus_manager_test.cpp [hunk 12]
```diff
@@ -6,6 +6,298 @@
...
+          .name = "Unknown stop along the route",
+          .config = {
+              PostBusRequest{
+                  .bus = "Bus 1",
+                  .stops = {"stop 1", "stop 2", "stop 4", "stop 1"},
+              },
+              PostStopRequest{
+                  .stop = "stop 1",
+                  .coords = {11.111111, 11.111111},
+                  .stops = {},
```
Comment (foxfrikses) [inline tests/bus_manager_test.cpp:139]: I believe you can skip empty `.stops`

File: src/bus_manager.cpp (hunk 7 above)
Comment (siarheishut) [reply src/bus_manager.cpp:42]: Changed it to ```string_view```.

File: tests/bus_manager_test.cpp (hunk 12 above)
Comment (siarheishut) [reply tests/bus_manager_test.cpp:139]: Ok, removed it.

File: tests/bus_manager_test.cpp (hunk 11 above)
Comment (siarheishut) [reply tests/bus_manager_test.cpp:296]: Changed it.

File: tests/bus_manager_test.cpp (hunk 8 above)
Comment (siarheishut) [reply tests/bus_manager_test.cpp:370]: Done.

File: tests/bus_manager_test.cpp (hunk 9 above)
Comment (siarheishut) [reply tests/bus_manager_test.cpp:462]: Done.

File: tests/bus_manager_test.cpp (hunk 10 above)
Comment (siarheishut) [reply tests/bus_manager_test.cpp:470]: Done.

File: src/bus_manager.cpp [hunk 13]
```diff
@@ -37,6 +38,32 @@ int ComputeUniqueCount(std::vector<std::string> stops) {
 }
 
 namespace rm {
+std::unique_ptr<BusManager> BusManager::Create(std::vector<PostRequest> &requests) {
```
Comment (foxfrikses) [inline src/bus_manager.cpp:41]: When you pass a param to a function via a non-const ref, it implies that this object is input-output param. Why do you make it non-const ref?

Comment (siarheishut) [reply src/bus_manager.cpp:41]: Changed to const ref.

**Review by foxfrikses — APPROVED** [review]
LGTM

**Pull request merged**